### ISSUE TITLE
Error handling on ldap sync

### ISSUE THREAD

**Issue by hartois** [glpi-project/glpi#984]
GLPI Version: 9.1

Some times I see warnings in php-errors.log:
  **\* PHP Warning(2): ldap_search(): Search: Can't contact LDAP server
  Backtrace :
  :  
  inc/authldap.class.php:2481                        ldap_search()
  inc/authldap.class.php:2068                        AuthLDAP::searchUserDn()
  scripts/ldap_mass_sync.php:164                     AuthLDAP::ldapImportUserByServerId()
  scripts/ldap_mass_sync.php:119                     import()

Maybe it's due to the server load, network problems or something else...

Today at the same time my 512 users have been Disabled.

I think that ldap sync does not handling error on LDAP operations.
And unsuccessful operations equal to the user have been removed from LDAP.

Second question: does ldap sync must Enable user that have been previous Disabled?

**Comment by hartois** [comment]
$infos = self::searchUserDn($ds,$attribs);         <------- return false

if ($infos && $infos['dn']) {                                     <--------- skipped by $infos == false
...
}

if ($action != self::ACTION_IMPORT) {            <--------- $action = self::ACTION_ALL, entering if statement
        $users_id = User::getIdByField('name', $params['value']);
        User::manageDeletedUserInLdap($users_id);
        return array('action' => self::USER_DELETED_LDAP,
                      'id'    => $users_id);
}
Except in this case, the user will not be perceived as a deleted?

On the second question.
If the user has been deleted in the ldap and disabled in glpi by ldap sync, then after it is re-added to the ldap - it is not enabled in glpi.

**Comment by trasher** [comment]
OK, I think I understand the issue. When the user is not found, searchuserDn() returns false, and then user will be deleted at one point.
But when this same method fails, it also returns false, and user is deleted even if it still exists (can't know beacause ldap did not respond).

As for the second question, I do not know if this is the expected behavior, I'll check that.

**Comment by hartois** [comment]
You understand correctly.

For the second question - I also do not know, it is expected behavior or not. 
I think - this should be an option.

**Comment by remicollet** [comment]
> If the user has been deleted in the ldap and disabled in glpi by ldap sync,

This is an optional behavior from configuration (keep, disable, delete, remove rights)

> then after it is re-added to the ldap  it is not enabled in glpi.

Indeed (excepted if you use "remove rights")

**Comment by trasher** [comment]
Thanks @remicollet for the clarifications ;)

**Comment by hartois** [comment]
> this should probably be open as another issue (change proposal)
1. I must open issue?
2. You agree that this behavior should be an option? Or you think that it should be fixed behavior for fixed sync behavior (keep, disable, delete, remove rights)?

**Comment by remicollet** [comment]
@hartois I mean that the "re-enable" option should be discussed, and is not related to this issue (related to sync)

**Comment by trasher** [comment]
OK, for the disconnect issue between connect and search; I'll have to get deep in the project, and recrreate locally an ldap with data I can change... 
Well, that is a quite huge job for now; that will not be fixed before the final 9.1 is out I guess (but this does not seem 9.1 related anyways; you should have the same behavior with older releases.)

**Comment by hartois** [comment]
I think - this is an old bug.
The more users and slower network (or than ldap farther from glpi ), the better the chance encounter with  it

**Comment by hartois** [comment]
Until this bug will not be fixed, I will decide the issue by using the following patch:
```
Index: inc/authldap.class.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- inc/authldap.class.php      (revision 5bd42841159cfc7d16ff8a4d7686bd2139ba5285)
+++ inc/authldap.class.php      (revision )
@@ -43,6 +43,8 @@
    const SIMPLE_INTERFACE = 'simple';
    const EXPERT_INTERFACE = 'expert';
 
+   const OPERATION_FAILURE  = -1;
+
    const ACTION_IMPORT      = 0;
    const ACTION_SYNCHRONIZE = 1;
    const ACTION_ALL         = 2;
@@ -2067,6 +2069,9 @@
 
          $infos = self::searchUserDn($ds,$attribs);
 
+         if ($infos === self::OPERATION_FAILURE)
+            return false;
+
          if ($infos && $infos['dn']) {
             $user_dn = $infos['dn'];
             $login   = $infos[$config_ldap->fields['login_field']];
@@ -2484,9 +2489,11 @@
          if (is_array($info) && ($info['count'] == 1)) {
             return array('dn'        => $info[0]['dn'],
                          $login_attr => $info[0][$login_attr][0]);
+         }else{
+            return false;
          }
       }
-      return false;
+      return self::OPERATION_FAILURE;
    }
```

**Comment by trasher** [comment]
Should be fixed in 9.2 ;)
